internal/access/grant.go

```diff
@@ -20,58 +20,36 @@ func GetGrant(c *gin.Context, id uid.ID) (*models.Grant, error) {
 }
 
 func ListGrants(c *gin.Context, subject uid.PolymorphicID, resource string, privilege string, inherited bool, showSystem bool, p *data.Pagination) ([]models.Grant, error) {
-	selectors := []data.SelectorFunc{
-		data.ByOptionalResource(resource),
-		data.ByOptionalPrivilege(privilege),
-	}
-
+	rCtx := GetRequestContext(c)
 	roles := []string{models.InfraAdminRole, models.InfraViewRole, models.InfraConnectorRole}
-	db, err := RequireInfraRole(c, roles...)
+	_, err := RequireInfraRole(c, roles...)
 	err = HandleAuthErr(err, "grants", "list", roles...)
 	if errors.Is(err, ErrNotAuthorized) {
 		// Allow an authenticated identity to view their own grants
-		db = getDB(c)
-		subjectID, err2 := subject.ID()
-		if err2 != nil {
-			// user is only allowed to select their own grants, so if the subject is missing or invalid, this is an access error
-			return nil, err
-		}
-		identity := AuthenticatedIdentity(c)
+		subjectID, _ := subject.ID() // zero value will never match a user
 		switch {
-		case identity == nil:
+		case rCtx.Authenticated.User == nil:
 			return nil, err
-		case subject.IsIdentity() && identity.ID == subjectID:
-			if inherited {
-				selectors = append(selectors, data.GrantsInheritedBySubject(subject))
-			} else {
-				selectors = append(selectors, data.BySubject(subject))
-			}
-			return data.ListGrants(db, p, selectors...)
-		case subject.IsGroup() && userInGroup(db, identity.ID, subjectID):
-			if inherited {
-				selectors = append(selectors, data.GrantsInheritedBySubject(subject))
-			} else {
-				selectors = append(selectors, data.BySubject(subject))
-			}
-			return data.ListGrants(db, p, selectors...)
+		case subject.IsIdentity() && rCtx.Authenticated.User.ID == subjectID:
+			// authorized because the request is for their own grants
+		case subject.IsGroup() && userInGroup(rCtx.DBTxn, rCtx.Authenticated.User.ID, subjectID):
+			// authorized because the request is for grants of a group they belong to
```

Comment:
This function did not need to perform the query in different places. Once the authorization check is complete we can perform the query once.

These checks here ensure that if the user does not have the necessary role, the query contains their own userID or a groupID they belong to. The `subject` here is the query parameter.